Sources/X509/Verifier/RFC5280/NameConstraintsPolicy.swift

```diff
@@ -21,7 +21,224 @@ struct NameConstraintsPolicy: VerifierPolicy {
 
     @inlinable
     func chainMeetsPolicyRequirements(chain: UnverifiedCertificateChain) -> PolicyEvaluationResult {
-        // For now we don't implement this, we'll add support later.
+        // The rules for name constraints come from https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.10.
+        //
+        // Some notes:
+        //
+        // - RFC 5280 says we MUST validate directoryName constraints, and SHOULD validate rfc822Name,
+        //       URI, dNSName, and iPAddress constraints.
+        // - If there's a constraint we don't support and can't validate, we MUST reject the cert.
+        //
+        // Our algorithm is recursive: starting from the root and moving towards the leaf, for each CA
+        // cert we apply the name constraints to all of the other certificates in the chain. The one exception
+        // is for self-signed certs where, much like with basic constraints, we briefly pretend that the
+        // self-signed cert issued itself and enforce its own name constraints on it.
+        if chain.count == 1 {
+            return Self._validateNameConstraints(chain[...], issuer: chain.first!)
+        }
+
+        var issuedCerts = chain[...]
+        while let issuer = issuedCerts.popLast(), issuedCerts.count > 0 {
```

Comment:
Took me a hot second to understand this. Clever and performant use of `ArraySlice` though.